README.md
# CRCCalc README
Simple plugin to calculate the CRC of selected text.
## Features

- 32 bit CRC algorithms (BZIP2, C, D, MPEG-2, POSIX, Q, JAMCRC, XFER)
- 16 bit CRC algorithms (CCITT-FALSE, ARC, 0xBB3D, AUG-CCITT, BUYPASS, CDMA2000, DDS-110, DECT-R, DECT-X, DNP, EN-13757, GENIBUS, MAXIM, MCRF4XX, RIELLO, T10-DIF, TELEDISK, TMS37157, USB, A, KERMIT, MODBUS, X-25, XMODEM)
- 8 bit CRC algorithms (CDMA2000, DARC, DVB-S2, EBU, I-CODE, ITU, MAXIM. ROHX, WCDMA)

![Screenshot](images/screenshot1.gif "Convert text to CRC")
## Release Notes

### 0.1.1

Initial release

-----------------------------------------------------------------------------------------------------------

**Enjoy!**


CHANGELOG.md
# Change Log
## [0.1.1] - 2020-04-21
### Added
- Initial release


src/crc.ts
//------------------------------------------------------------------------------
// MIT License
//
// Copyright (c) 2021 Tobias Barendt
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//------------------------------------------------------------------------------

class CRC
{
	//--------------------------------------------------------------------------
	// Constructor
	//--------------------------------------------------------------------------
	constructor(bits: number, poly: number, init: number, reflIn: boolean, reflOut: boolean, xor: number, name: string)
	{
		// Cast & msb masks
		if(bits == 8)
			this.m_castMask = 0xFF;
		else if(bits == 16)
			this.m_castMask = 0xFFFF;
		else if(bits == 32)
			this.m_castMask = 0xFFFFFFFF;
		else
			throw "Invalid bit length, must be 8, 16 or 32!";
		var msbMask = 0x01 << (bits - 1);

		// Cache settings
		this.m_name = name;
		this.m_bits = bits;
		this.m_initialValue = init;
		this.m_xor = xor;
		this.m_reflIn = reflIn;
		this.m_reflOut = reflOut;

		// Create table
		this.m_crcTable = new Array(256);
		for(var i = 0; i < 256; i++)
		{
			var byte = (i << (this.m_bits - 8)) & this.m_castMask;
			for(var bit = 0; bit < 8; bit++)
			{
				if((byte & msbMask) != 0)
				{
					byte <<= 1;
					byte ^= poly;
				}
				else
				{
					byte <<= 1;
				}
			}
			this.m_crcTable[i] = (byte & this.m_castMask);
		}
	}

	//--------------------------------------------------------------------------
	// Compute
	//--------------------------------------------------------------------------
	public Compute(str: string)
	{
		var crc = this.m_initialValue;
		for(var i = 0; i < str.length; i++)
		{
			var byte = str.charCodeAt(i);
			if(this.m_reflIn)
				byte = this.Reflect(8, byte);
			
			crc = (crc ^ (byte << (this.m_bits - 8))) & this.m_castMask; 
			var t = (crc >> (this.m_bits - 8)) & 0xFF;
			crc = (crc << 8) & this.m_castMask;
			crc = (crc ^ this.m_crcTable[t]) & this.m_castMask;
		}

		if(this.m_reflOut)
			crc = this.Reflect(this.m_bits, crc);
		crc ^= this.m_xor;
		crc &= this.m_castMask;
		return crc;
	}

	//--------------------------------------------------------------------------
	// Reflect
	//--------------------------------------------------------------------------
	private Reflect(bits : number, value : number) 
	{
		var reflected = 0;
		for(var i = 0; i < bits; i++)
		{
			if((value & (1 << i)) != 0)
			{
				reflected |= (1 << ((bits - 1) - i));
			}
		}
		return reflected;
	}

	//--------------------------------------------------------------------------
	// Name
	//--------------------------------------------------------------------------
	public get name(){return this.m_name;}

	//--------------------------------------------------------------------------
	// Bits
	//--------------------------------------------------------------------------
	public get Bits(){return this.m_bits;}

	//--------------------------------------------------------------------------
	// Settings
	private m_name: string = "";
	private m_bits: number = 0;
	private m_initialValue: number = 0;
	private m_xor: number = 0;
	private m_reflIn: boolean = false;
	private m_reflOut: boolean = false;
	private m_crcTable: number[] = [0];
	private m_castMask: number = 0;
}

//------------------------------------------------------------------------------
// Exports
export default CRC;


src/extension.ts
//------------------------------------------------------------------------------
// MIT License
//
// Copyright (c) 2021 Tobias Barendt
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//------------------------------------------------------------------------------

import * as vscode from 'vscode';
import CRC from './crc';

var listOfCRC =
	[
		new CRC(32, 0x04C11DB7, 0xFFFFFFFF, true, true, 0xFFFFFFFF, "CRC-32"),
		new CRC(32, 0x04C11DB7, 0xFFFFFFFF, false, false, 0xFFFFFFFF, "CRC-32/BZIP2"),
		new CRC(32, 0x1EDC6F41, 0xFFFFFFFF, true, true, 0xFFFFFFFF, "CRC-32C"),
		new CRC(32, 0xA833982B, 0xFFFFFFFF, true, true, 0xFFFFFFFF, "CRC-32D"),
		new CRC(32, 0x04C11DB7, 0xFFFFFFFF, false, false, 0x00000000, "CRC-32/MPEG-2"),
		new CRC(32, 0x04C11DB7, 0x00000000, false, false, 0xFFFFFFFF, "CRC-32/POSIX"),
		new CRC(32, 0x814141AB, 0x00000000, false, false, 0x00000000, "CRC-32Q"),
		new CRC(32, 0x04C11DB7, 0xFFFFFFFF, true, true, 0x00000000, "CRC-32/JAMCRC"),
		new CRC(32, 0x000000AF, 0x00000000, false, false, 0x00000000, "CRC-32/XFER"),

		new CRC(16, 0x8005, 0x0000, true, true, 0x0000, "CRC-16/ARC"),
		new CRC(16, 0x1021, 0xFFFF, false, false, 0x0000, "CRC-16/CCITT-FALSE"),
		new CRC(16, 0x1021, 0x1D0F, false, false, 0x0000, "CRC-16/AUG-CCITT"),
		new CRC(16, 0x8005, 0x0000, false, false, 0x0000, "CRC-16/BUYPASS"),
		new CRC(16, 0xC867, 0xFFFF, false, false, 0x0000, "CRC-16/CDMA2000"),
		new CRC(16, 0x8005, 0x800D, false, false, 0x0000, "CRC-16/DDS-110"),
		new CRC(16, 0x0589, 0x0000, false, false, 0x0001, "CRC-16/DECT-R"),
		new CRC(16, 0x0589, 0x0000, false, false, 0x0000, "CRC-16/DECT-X"),
		new CRC(16, 0x3D65, 0x0000, true, true, 0xFFFF, "CRC-16/DNP"),
		new CRC(16, 0x3D65, 0x0000, false, false, 0xFFFF, "CRC-16/EN-13757"),
		new CRC(16, 0x1021, 0xFFFF, false, false, 0xFFFF, "CRC-16/GENIBUS"),
		new CRC(16, 0x8005, 0x0000, true, true, 0xFFFF, "CRC-16/MAXIM"),
		new CRC(16, 0x1021, 0xFFFF, true, true, 0x0000, "CRC-16/MCRF4XX"),
		new CRC(16, 0x1021, 0xB2AA, true, true, 0x0000, "CRC-16/RIELLO"),
		new CRC(16, 0x8BB7, 0x0000, false, false, 0x0000, "CRC-16/T10-DIF"),
		new CRC(16, 0xA097, 0x0000, false, false, 0x0000, "CRC-16/TELEDISK"),
		new CRC(16, 0x1021, 0x89EC, true, true, 0x0000, "CRC-16/TMS37157"),
		new CRC(16, 0x8005, 0xFFFF, true, true, 0xFFFF, "CRC-16/USB"),
		new CRC(16, 0x1021, 0xC6C6, true, true, 0x0000, "CRC-A"),
		new CRC(16, 0x1021, 0x0000, true, true, 0x0000, "CRC-16/KERMIT"),
		new CRC(16, 0x8005, 0xFFFF, true, true, 0x0000, "CRC-16/MODBUS"),
		new CRC(16, 0x1021, 0xFFFF, true, true, 0xFFFF, "CRC-16/X-25"),
		new CRC(16, 0x1021, 0x0000, false, false, 0x0000, "CRC-16/XMODEM"),
		
		new CRC(8, 0x07, 0x00, false, false, 0x00, "CRC-8"),
		new CRC(8, 0x9B, 0xFF ,false ,false, 0x00, "CRC-8/CDMA2000"),
		new CRC(8, 0x39, 0x00, true, true, 0x00, "CRC-8/DARC"),
		new CRC(8, 0xD5, 0x00 ,false, false, 0x00, "CRC-8/DVB-S2"),
		new CRC(8, 0x1D, 0xFF, true, true, 0x00, "CRC-8/EBU"),
		new CRC(8, 0x1D, 0xFD, false, false, 0x00, "CRC-8/I-CODE"),
		new CRC(8, 0x07, 0x00, false, false, 0x55, "CRC-8/ITU"),
		new CRC(8, 0x31, 0x00, true, true, 0x00, "CRC-8/MAXIM"),
		new CRC(8, 0x07, 0xFF, true, true, 0x00, "CRC-8/ROHC"),
		new CRC(8, 0x9B, 0x00, true, true, 0x00, "CRC-8/WCDMA"),
	];


//------------------------------------------------------------------------------
// activate
//------------------------------------------------------------------------------
export function activate(context: vscode.ExtensionContext)
{
	for(var i = 0; i < listOfCRC.length; i++)
	{
		let crcTable = listOfCRC[i];
		const cmdFunc = () =>
		{
			var editor = vscode.window.activeTextEditor;
			if(!editor)
				return;

			editor.selections.forEach((selection) =>
			{
				var editor = vscode.window.activeTextEditor;
				if(!editor)
					return;
				editor.edit((selectedText) =>
				{
					var editor = vscode.window.activeTextEditor;
					if(!editor)
						return;
					var doc = editor.document;
					var text = doc.getText(selection);
					function PadString(str: string, minLength: number, symbol: string)
					{
						var zero = 1 + minLength - str.toString().length;
						return Array(+(zero > 0 && zero)).join(symbol) + str;
					}

					
					let crc = crcTable.Compute(text);
					if(crcTable.Bits == 32)
					{
						let crc1 = (crc & 0xFFFF0000) >>> 16;
						let crc2 = crc & 0xFFFF;
						selectedText.replace(selection, "0x" + PadString(crc1.toString(16), 4, "0").toUpperCase() + PadString(crc2.toString(16), 4, "0").toUpperCase() + " /*" + text + "*/");
					}
					else if(crcTable.Bits == 16)
					{
						selectedText.replace(selection, "0x" + PadString(crc.toString(16), 4, "0").toUpperCase() + " /*" + text + "*/");
					}
					else if(crcTable.Bits == 8)
					{
						selectedText.replace(selection, "0x" + PadString(crc.toString(16), 2, "0").toUpperCase() + " /*" + text + "*/");
					}
				});
			});
		}

		context.subscriptions.push(vscode.commands.registerCommand("crccalc." + crcTable.name, cmdFunc));
	}
}

//------------------------------------------------------------------------------
// Exports
export function deactivate() {}
